wp-content/themes/core/components/tabs/Tabs_Controller.php

```diff
@@ -0,0 +1,244 @@
+		$args = $this->parse_args( $args );
+
+		$this->tabs               = (array) $args[ self::TABS ];
+		$this->classes            = (array) $args[ self::CLASSES ];
+		$this->attrs              = (array) $args[ self::ATTRS ];
+		$this->layout             = (string) $args[ self::LAYOUT ];
+		$this->toggle_classes     = (array) $args[ self::TOGGLE_CLASSES ];
+		$this->tab_button_classes = (array) $args[ self::TAB_BUTTON_CLASSES ];
+		$this->tab_panel_classes  = (array) $args[ self::TAB_PANEL_CLASSES ];
+		$this->tablist_id         = uniqid( 'c-tabs__tablist--' );
+
+		$this->init_tabs();
+	}
+
+	/**
+	 * @return array
+	 */
+	protected function defaults(): array {
+		return [
+			self::TABS               => [],
+			self::CLASSES            => [],
+			self::ATTRS              => [],
+			self::LAYOUT             => self::LAYOUT_HORIZONTAL,
+			self::TOGGLE_CLASSES     => [],
+			self::TAB_BUTTON_CLASSES => [],
+			self::TAB_PANEL_CLASSES  => [ 's-sink', 't-sink' ],
+		];
+	}
+
+	/**
+	 * @return array
+	 */
+	protected function required(): array {
+		return [
+			self::CLASSES            => [ 'c-tabs' ],
+			self::ATTRS              => [ 'data-js' => 'c-tabs' ],
+			self::TOGGLE_CLASSES     => [ 'c-tabs__tablist-toggle' ],
+			self::TAB_BUTTON_CLASSES => [ 'c-tabs__tab' ],
+			self::TAB_PANEL_CLASSES  => [ 'c-tabs__tabpanel' ],
+		];
+	}
+
+	/**
+	 * Loop through the tabs provided setup the tab button and tab panel components arguments for each.
+	 */
+	private function init_tabs() {
+		foreach ( $this->tabs as $index => $tab ) {
+			$tab_id              = uniqid();
+			$this->tab_buttons[] = $this->get_tab_button_args( $tab, $tab_id, $index );
+			$this->tab_panels[]  = $this->get_tab_panel_args( $tab, $tab_id, $index );
+		}
+	}
+
+	/**
+	 * Return arguments for an individual tab list "tab" button.
+	 *
+	 * @param Tab_Model $tab
+	 * @param string    $tab_id
+	 * @param int       $index
+	 *
+	 * @return array
+	 */
+	private function get_tab_button_args( Tab_Model $tab, string $tab_id, int $index ): array {
+		$args = [
+			'content' => $tab->label ?: sprintf( __( 'Tab %d', 'tribe' ), $index + 1 ),
+			'classes' => $this->tab_button_classes,
+			'attrs'   => [
+				'id'            => sprintf( 'c-tabs__tab--%s', $tab_id ),
+				'aria-controls' => sprintf( 'c-tabs__tabpanel--%s', $tab_id ),
+				'aria-selected' => $index === 0 ? 'true' : 'false',
+				'data-js'       => 'c-tabs__button',
+				'role'          => 'tab',
+			],
+		];
+
+		if ( $index !== 0 ) {
+			$args['attrs']['tabindex'] = '-1';
+		}
+
+		return $args;
+	}
+
+	/**
+	 * Return arguments for an individual tab content, the "tabpanel".
+	 *
+	 * @param Tab_Model $tab
+	 * @param string    $tab_id
+	 * @param int       $index
+	 *
+	 * @return array
+	 */
+	private function get_tab_panel_args( Tab_Model $tab, string $tab_id, int $index ): array {
+		$args = [
+			'content' => $tab->content ?: '&nbsp;', // If the tab content is empty, the container component won't render.
```

Comment:
same comment as above, escaping on `$tab->content`?

Reply:
This for sure will contain html content, so similar question to above: what should be escaped versus what should be allowed?